src/neptune/main.d

```diff
@@ -0,0 +1,342 @@
+
+    Structure to define a command to be run locally along with a description of
+    it
+
+*******************************************************************************/
+
+struct LocalAction
+{
+    string command;
+    string description;
+}
+
+/*******************************************************************************
+
+    Structure to collect actions and metatadata about those actions
+
+*******************************************************************************/
+
+struct ActionList
+{
+    /// List of actions to execute, local only
+    LocalAction[] local_actions;
+
+    /// tags/branches that were locally modified and will be pushed
+    string[] affected_refs;
+
+    /// Releases to do on github
+    string[] github_releases;
+
+    ref ActionList opOpAssign ( string op ) ( const ActionList list )
+    {
+        this.local_actions ~= list.local_actions;
+        this.affected_refs ~= list.affected_refs;
+        this.github_releases ~= list.github_releases;
+
+        return this;
+    }
+}
+
+/*******************************************************************************
+
+    Programs entry point
+
+*******************************************************************************/
+
+void main ( )
+{
+    import std.stdio;
+    import std.algorithm : map, sort;
+    import std.range : array;
+
+    //import vibe.core.log;
+    //setLogLevel(LogLevel.trace);
+
+    auto repo = api.repository(getUpstream());
+    auto tags = repo.releasedTags().map!(a=>Version(a.name)).array;
+
+    sort(tags);
+
+    cmd("git remote update");
+
+    auto release = autodetect(tags);
+
+    ActionList list;
+
+    if (release.type == release.type.Patch)
+        list = preparePatchRelease(repo, tags, release);
+    else
+    {
+        writefln("This seems to be %s release which is not yet supported, exiting...",
+                 release.type);
+        return;
+    }
+
+    writefln("Actions to be done:");
+
+    foreach (action; list.local_actions)
+        writefln(" * %s (%s)", action.description, action.command);
+
+    writef("\nWould you like to continue? y/n: ");
+
+    if (!readYesNoResponse())
+    {
+        writefln("Aborting...");
+        return;
+    }
+
+
+    writefln("");
+    foreach (action; list.local_actions)
+    {
+        import std.string : strip;
+
+        writefln("%s ...\n\t%s",
+                 action.description,
+                 action.command);
+        writefln("\t%s", strip(cmd(action.command)));
+    }
+
+
+    writefln("Done! Would you like to push the modified references now?");
+    writefln("References to be pushed: %s", list.affected_refs);
+    writef("y/n?: ");
+
+    if (!readYesNoResponse())
+    {
+        writefln("Aborting...");
+        return;
+    }
+
+    foreach (reference; list.affected_refs)
+    {
+        import std.format;
+
+        writefln("Pushing %s ...", reference);
+        cmd(format("git push %s %s", getRemote(), reference));
```

Comment:
That's not as easy because I don't have the branch/tag info anymore at that point.
I guess I _could_ guess it from the name